### ISSUE TITLE
Installation [fstcore/ZSTD/decompress/huf_decompress.o] error

### ISSUE THREAD

**Issue by dweemx** [fstpackage/fst#167]
Hi,

I'm trying to install fst R package but I'm getting this error:
```
gcc -std=gnu99 -I$HOME/software/R/3.4.2-X11/lib64/R/include -DNDEBUG -fopenmp -I. -Ifstcore -Ifstcore_v1 -Ifstcore/LZ4 -Ifstcore/ZSTD -Ifstcore/ZSTD/common -Ifstcore/ZSTD/decompress -Ifstcore/ZSTD/compress -I"$HOME/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I $HOME/software/libreadline/6.3-foss-2014a/include/ -I $HOME/software/zlib/1.2.8-foss-2014a/include/ -I $HOME/software/bzip2/1.0.6-foss-2014a/include/ -I $HOME/software/XZ/5.0.5-foss-2014a/include/ -I $HOME/software/biomed/PCRE/8.39-foss-2014a/include/ -I $HOME/software/cURL/7.34.0-foss-$HOME/opt/xorg/include/   -fpic  -g -O2 
-c fstcore/ZSTD/decompress/huf_decompress.c -o fstcore/ZSTD/decompress/huf_decompress.o
/tmp/ccFKeDcT.s: Assembler messages:
/tmp/ccFKeDcT.s:1593: Error: no such instruction: `shrx %rsi,%r8,%r8'
/tmp/ccFKeDcT.s:1628: Error: no such instruction: `shrx %rsi,%r8,%r8'
/tmp/ccFKeDcT.s:1663: Error: no such instruction: `shrx %rsi,%r8,%r8'
/tmp/ccFKeDcT.s:1706: Error: no such instruction: `shrx %rsi,%rdx,%rdx'
/tmp/ccFKeDcT.s:1781: Error: no such instruction: `shrx %r10,%rdx,%rdx'
/tmp/ccFKeDcT.s:2078: Error: no such instruction: `shrx %rsi,%rdx,%rdx'
/tmp/ccFKeDcT.s:3587: Error: no such instruction: `shrx %rsi,%r13,%r13'
/tmp/ccFKeDcT.s:3625: Error: no such instruction: `shrx %rsi,%r13,%r13'
/tmp/ccFKeDcT.s:3663: Error: no such instruction: `shrx %rsi,%r13,%r13'
/tmp/ccFKeDcT.s:3690: Error: no such instruction: `shrx %rsi,%r9,%r9'
/tmp/ccFKeDcT.s:3765: Error: no such instruction: `shrx %rax,%rsi,%rsi'
/tmp/ccFKeDcT.s:5384: Error: no such instruction: `shrx %r11,%rdx,%rdx'
/tmp/ccFKeDcT.s:5422: Error: no such instruction: `shrx %r11,%rdx,%rdx'
/tmp/ccFKeDcT.s:5467: Error: no such instruction: `shrx %r11,%r8,%r8'
/tmp/ccFKeDcT.s:5515: Error: no such instruction: `shrx %r11,%r10,%rcx'
/tmp/ccFKeDcT.s:5558: Error: no such instruction: `shrx %r11,%r10,%rcx'
/tmp/ccFKeDcT.s:5614: Error: no such instruction: `shrx %r11,%r10,%rcx'
/tmp/ccFKeDcT.s:5657: Error: no such instruction: `shrx %r11,%r9,%rcx'
/tmp/ccFKeDcT.s:5698: Error: no such instruction: `shrx %r11,%r14,%rcx'
/tmp/ccFKeDcT.s:5760: Error: no such instruction: `shrx %r11,%r14,%rcx'
/tmp/ccFKeDcT.s:5815: Error: no such instruction: `shrx %r11,%rdi,%rcx'
/tmp/ccFKeDcT.s:5871: Error: no such instruction: `shrx %r11,%rsi,%rcx'
/tmp/ccFKeDcT.s:5928: Error: no such instruction: `shrx %r11,%rdx,%rdx'
/tmp/ccFKeDcT.s:5964: Error: no such instruction: `shrx %r11,%rax,%rax'
/tmp/ccFKeDcT.s:6008: Error: no such instruction: `shrx %r11,%r14,%rcx'
/tmp/ccFKeDcT.s:6069: Error: no such instruction: `shrx %r11,%r10,%rcx'
/tmp/ccFKeDcT.s:6134: Error: no such instruction: `shrx %r11,%r10,%rcx'
/tmp/ccFKeDcT.s:9806: Error: no such instruction: `shrx %r8,%rax,%rax'
/tmp/ccFKeDcT.s:9842: Error: no such instruction: `shrx %r8,%rax,%rax'
/tmp/ccFKeDcT.s:9887: Error: no such instruction: `shrx %r8,%r11,%rcx'
/tmp/ccFKeDcT.s:9933: Error: no such instruction: `shrx %r8,%r11,%rcx'
/tmp/ccFKeDcT.s:9968: Error: no such instruction: `shrx %r8,%r15,%rcx'
/tmp/ccFKeDcT.s:10003: Error: no such instruction: `shrx %r8,%r15,%rcx'
/tmp/ccFKeDcT.s:10038: Error: no such instruction: `shrx %r8,%r15,%rcx'
/tmp/ccFKeDcT.s:10078: Error: no such instruction: `shrx %r8,%r15,%rcx'
/tmp/ccFKeDcT.s:10113: Error: no such instruction: `shrx %r8,%r15,%rcx'
/tmp/ccFKeDcT.s:10148: Error: no such instruction: `shrx %r8,%r15,%rcx'
/tmp/ccFKeDcT.s:10183: Error: no such instruction: `shrx %r8,%r15,%rcx'
/tmp/ccFKeDcT.s:10223: Error: no such instruction: `shrx %r8,%r15,%rcx'
/tmp/ccFKeDcT.s:10265: Error: no such instruction: `shrx %r8,%rsi,%rcx'
/tmp/ccFKeDcT.s:10326: Error: no such instruction: `shrx %r8,%rdx,%rdx'
/tmp/ccFKeDcT.s:10392: Error: no such instruction: `shrx %r8,%rax,%rax'
/tmp/ccFKeDcT.s:10438: Error: no such instruction: `shrx %r8,%rax,%rax'
make: *** [fstcore/ZSTD/decompress/huf_decompress.o] Error 1
ERROR: compilation failed for package ‘fst’
```

I'm compiling with GCC verison 4.8.2

Any idea what's going wrong?

**Comment by MarcusKlik** [comment]
Hi @mase5, thanks for reporting!

It looks like your build fails on the `ZSTD` (compressor) component. I'm not sure what the minimum `gcc` version is for building `ZSTD`, but I can see from the [CRAN build results](https://cran.r-project.org/web/checks/check_results_fst.html) that they use version 4.9.3 or higher.

Are you able to update your version of the `gcc` compiler ?

thanks!

**Comment by rushgeo** [comment]
This is happening to me on a Centos 6.4 cluster, but the package installs fine on every Ubuntu 16.04 system I've tried.  

Since the cluster uses environment modules, I've tried the following combinations, but all have failed on `huf_decompress.o`:

- R 3.3.1 and gcc 4.8.2
- R 3.5.1 and gcc 5.3.0
- R 3.5.1 and gcc 7.3.0

My error log looks similar:
```
gcc -std=gnu99 -I"/hpc/packages/minerva-common/R/3.5.1/lib64/R/include" -DNDEBUG  -I. -Ifstcore -Ifstcore_v1 -Ifstcore/LZ4 -Ifstcore/ZSTD -Ifstcore/ZSTD/common -Ifstcore/ZSTD/decompress -Ifstcore/ZSTD/compress -I"/hpc/packages/minerva-common/rpackages/3.5.1/site-library/Rcpp/include" -I/usr/local/include   -fpic  -O3 -fopenmp  -c fstcore/ZSTD/decompress/huf_decompress.c -o fstcore/ZSTD/decompress/huf_decompress.o
/tmp/ccdufUvI.s: Assembler messages:
/tmp/ccdufUvI.s:212: Error: no such instruction: `shrx %eax,(%r8),%r10d'
<... 89 more lines of no such instruction ...>
make: *** [fstcore/ZSTD/decompress/huf_decompress.o] Error 1
ERROR: compilation failed for package ‘fst’
```

The only difference in the versions I've tried is that for R 3.3.1, there were only 44 total missing assembler messages instead of 90 for R 3.5.1. Let me know if you want the complete list of all of those.

Is there anything else I can test to help chase this down?

**Comment by MarcusKlik** [comment]
Hi @rushgeo, thanks for testing and for trying to get `fst` to work on your `Centos` cluster!

From similar issues reported on `Centos 6` (for example with the [`hyperscan`](https://github.com/intel/hyperscan) library, [this](https://github.com/intel/hyperscan/issues/44) and [this](https://github.com/intel/hyperscan/issues/53) issue), it looks like there is a mismatch between the compiler and assembler, e.g. the compiler is generating `SIMD` instructions which are not understood by the assembler used.

We could test this assumption by disabling vectorization optimization of the `gcc` compiler. Perhaps you could try changing the top line in `fst`'s _src/Makevars_ file into:

```bash
PKG_CPPFLAGS = $(SHLIB_OPENMP_CXXFLAGS) -I. -Ifstcore -Ifstcore_v1 -Ifstcore/LZ4 \
        -Ifstcore/ZSTD -Ifstcore/ZSTD/common -Ifstcore/ZSTD/decompress -Ifstcore/ZSTD/compress \
        -fno-tree-vectorize
```

`gcc` option _-fno-tree-vectorize_ should block the compiler from generating any `SIMD` instructions. If `fst` does compile it that case, at least we would have pinpointed the problem.

The `shrx` instruction is from the [`bmi2`](https://docs.oracle.com/cd/E36784_01/html/E36859/gnyak.html) assembly set, you could also try selectively disabling that set by using _-march=native -mno-bmi2_ instead of _-fno-tree-vectorize_ in _Makevars_.

And there is also the possibility that updating the [_devtoolset collections_](https://www.softwarecollections.org/en/scls/rhscl/devtoolset-7/) as suggested in [this comment](https://github.com/intel/hyperscan/issues/53#issuecomment-300028980) will solve the problem, are you able to execute updates on your cluster? A newer assembler might understand the `SIMD` instructions generated by the compiler.  

Thanks for spending time on this!

**Comment by rushgeo** [comment]
You were right about the assembler! Thanks for researching those similar issues. 

The default `as` version on my cluster is 2.20.51.0.2. But the cluster does have an available environment module for `binutils` v2.29, and after loading that and a newer `gcc`, the install works.

For anyone else that might run into this on a cluster, here's what I ran:
```
module load R/3.5.1
module swap gcc gcc/7.3.0  # swap, because the R module on my system includes an older gcc module
module load binutils/2.29.1
R # then install the package
```

Other users will need to substitute different version numbers for whatever their available modules are.

**Comment by MarcusKlik** [comment]
Hi @rushgeo, that's great, I'm glad you got `fst` to compile on the cluster!

Thanks for sharing your solution!

**Comment by strazto** [comment]
@rushgeo thank you very much for the solution, it did help me get further in installation of the package on a cluster.

```bash
module load R/3.6.1
module swap gcc gcc/7.4.0
module load binutils/2.30
```

Unfortunately, I still hit an error in installation:

```
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘fst’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/home/mstr3336/software/fst/renv/library/R-3.6/x86_64-pc-linux-gnu/00LOCK-fst/00new/fst/libs/fst.so':
  /home/mstr3336/software/fst/renv/library/R-3.6/x86_64-pc-linux-gnu/00LOCK-fst/00new/fst/libs/fst.so: undefined symbol: _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEC1ERKS4_
Error: loading failed
Execution halted
ERROR: loading failed
```

**Comment by strazto** [comment]
This new build error is documented here: https://github.com/ContinuumIO/anaconda-issues/issues/3885

One person suggests that I need to use the same compiler that I ( cluster admins ) used to compile R.

To investigate this, I do the following:

```
R CMD config CFLAGS
```

```
-I/usr/local/hdf5/1.10.4/include -I/usr/local/boost/1.67.0/include -I/usr/local/jags/4.2.0/include -I/usr/local/protobuf/3.5.1/include -I/usr/local/tbb/4.4/include -I/usr/local/netcdf/4.6.1/include -I/usr/local/imagemagick/6.9.2-3/include -I/usr/local/nlopt/2.4.2/include -I/usr/local/geos/3.8.0/include -I/usr/local/jq/1.5/include -I/usr/local/cuda/10.0.130/include -L/usr/local/gsl/2.3/lib -I/usr/local/gsl/2.3/include -I/usr/local/lapack/3.8.0/include -I/usr/local/mysql/5.7.1/include -I/usr/local/sqlite/3260000/include -I/usr/local/proj/6.2.0/include -I/usr/local/udunits/2.2.20/include -I/usr/local/libtiff/4.0.3/include -I/usr/local/openssl/1.0.1s/include -I/usr/local/zlib/1.2.8/include -I/usr/local/pcre/8.39/include -I/usr/local/xz/5.2.2/include -I/usr/local/openmpi/3.1.3-gcc/include -I/usr/local/bzip2/1.0.6/include -I/usr/local/libcurl/7.40.0/include -I/usr/local/gcc/4.9.3/include -I/usr/local/mpfr/3.1.4/include -I/usr/local/gmp/5.1.3/include

```

I notice `-I/usr/local/gcc/4.9.3/include`

Which may imply that gcc/4.9.3 is what I need to use to compile `fst` on my system. 

I noticed a few other threads re. compilation errors, wherein @MarcusKlik states the minimum version of gcc required to compile `fst` .

https://github.com/fstpackage/fst/issues/116#issuecomment-358915488

> GCC with at least version 4.6 is needed to compile `fst`

https://github.com/fstpackage/fst/issues/185#issuecomment-449513089

> Indeed, it appears that _non-static data member initializers_ were introduced with [this `gcc` proposal](http://www.open-std.org/JTC1/SC22/WG21/docs/papers/2008/n2756.htm), which was implemented in `gcc` version 4.7 (see [this list](https://gcc.gnu.org/projects/cxx-status.html)).

The upshot is that the default version of gcc for my system, reported by `module avail` is: `gcc/4.9.3(default)`

I do note that this is actually the version of `gcc` that `module load R/3.6.1` load by default - I suspect that maybe the missing link is just `binutils/2.30`. 

I'll give this a try
